json-schema/json-schema/src/test/scala/endpoints4s/algebra/JsonSchemasDocs.scala

```diff
@@ -66,7 +66,7 @@ trait JsonSchemasDocs extends JsonSchemas {
   case class Recursive(next: Option[Recursive])
 
   val recursiveSchema: Record[Recursive] = (
-    optField("next")(lazyRecord(recursiveSchema, "Rec"))
+    optField("next")(lazySchema("Rec")(recursiveSchema))
   ).xmap(Recursive(_))(_.next)
```

Comment:
It would be nice to be able to write this:

~~~ scala
val recursiveSchema: Record[Recursive] =
  lazyRecord("Rec") {
    optField("next")(recursiveSchema)
  }.xmap(Recursive(_))(_.next)
~~~

But that does not work currently because `lazyRecord` returns a `JsonSchema` instead of a `Record`. Would it be possible to make it return a `Record` instead?

Reply:
I'm not sure I follow you. `lazySchema` (not `lazyRecord`) cannot return a `Record` because it doesn't always encode to a Json object. For instance with the `Expression` example, a simple literal value is encoded as a Json number.

As for `lazyRecord` and `lazyTagged`, they are deprecated. Are you suggesting that we add new versions of these of the following form?

```scala
def lazyRecord(name: String)(schema: => Record): Record
def lazyTagged(name: String)(schema: => Tagged): Tagged
```

Reply:
Yes, I realize now that we want a way to eventually return a `Record` or `Tagged`, like we could do in the `recursiveSchema` example. What do you think?

Reply:
Ok I've added these two methods. It was pretty easy in the `JsonSchemas` interpreters except for openapi. 

The problem there is that a `Record` is a class that expects a `DocumentedRecord` value in its constructor so the following doesn't work:

```scala
  override def lazyRecord[A](name: String)(schema: => Record[A]): Record[A] =
    new Record[A](
      ujsonSchemas.lazyRecord(name)(schema.ujsonSchema),
      LazySchema(name, schema.docs) // Wrong need a DocumentedRecord here
    )
```

So I made `DocumentedRecord` a `sealed abstract class` with two sub-classes. One is a simple `case class`, the other is a `Lazy` version. We also have to be careful about the other operations in the algebra that manipulate a `Record`.

For instance:

```scala
val recursiveSchema: Record[Recursive] =
  lazyRecord("Rec") {
    optField("next")(recursiveSchema)
  }
    .xmap(Recursive(_))(_.next)
    // We have to be careful that this doesn't result in a NPE
    .withTitle("Recursive is hard")
```

Reply:
What do you mean by “we have to be careful that this doesn’t result in a NPE”? Do users of endpoints4s have to be careful?

Reply:
Could you please clarify the above question?

Reply:
Ho sorry, I meant that we have to be careful in the implementation of the interpreter that it doesn't raise an NPE. This is not a concern for the end user.

In the tests, I added a couple of the `withXYZ` on the recursive schemas to make sure they work properly when they are used "insie" `lazyRecord` or on the result of it:

```scala
  val taggedRecursiveSchema: Tagged[TaggedRecursive] =
    lazyTagged("TaggedRec") {
      val aSchema = ...
      val bSchema = ...

      (aSchema.tagged("A") orElseMerge bSchema.tagged("B"))
        // Put some doc inside lazyTagged, and some more outside to make sure it works both way
        .withDiscriminator("kind")
        .withDescription("TaggedRec description")
    }
      .withTitle("TaggedRec title")
      .withExample(TaggedRecursiveA("foo", None))
```

Reply:
Awesome, thank you for the explanation!